### ISSUE TITLE
Possible data corruption during compaction

### ISSUE THREAD

**Issue by knizhnik** [neondatabase/neon#4295]
## Steps to reproduce

Compaction writes generated L1 layers to the disk and then calls `par_fsync::par_fsync` to flush them. But if pageserver node  is crashed at this moment then there is no warranty that  content of this L1 layers is correct after restart. 
Consistency of pageserver storage is mostly enforced b `disk_consistent_lsn`: we first flush changes , then advance `disk_consistent_lsn` and update metadata file. Or restart we remove all layers younger than `disk_consistent_lsn`.

But such approach doesn't work with L1 layers, because we are not advancing LSN here.
We need some other way to ensure L1 layer consistency.
The straightforward workaround is to perform fsync before renaming from temp file. 
But it may have negative impact on performance. We can try to combine rename and fsync - i.e. have somethig like `par_fsync::par_fsync_rename`  which will perform fsync+rename in parallel.

Alternatively we can calculate layer CRC and rely on it.

## Logs, links
-

**Comment by koivunej** [comment]
Additionally: With the fsyncs on the recent par_fsync changing #4265 I noted that we might fsync the parent directory before we fsync the files within the directory.

> But such approach doesn't work with L1 layers, because we are not advancing LSN here.

I think this is in relation to some tests as well because right now we sleep when hoping to compact and upload.

I have been thinking about a versioned identifier for the layer map, for example, disk_consitent_lsn + number of changes on top, each gc or compaction would be one change, so you'd have `lsn.0`, after compaction `lsn.1`, after gc `lsn.2`.

> We need some other way to ensure L1 layer consistency.
> The straightforward workaround is to perform fsync before renaming from temp file.
> But it may have negative impact on performance. We can try to combine rename and fsync - i.e. have somethig like `par_fsync::par_fsync_rename` which will perform fsync+rename in parallel.

I am not sure I agree here. 

For new L1s:
1. the DeltaFileWriter already does write to a temp file
2. fsync the temp file
3. then rename the file to it's real name
4. finally have the rename operation be fsynced (hopefully first for new layer file, *only then* to parent directory) later on lines: https://github.com/neondatabase/neon/blob/63884543758118b1e12b3654ae9bed57be570b39/pageserver/src/tenant/timeline.rs#L3421-L3430

So I think we do most of the things you consider here, but we don't have a way to make them visible because of no versioning.

**Comment by knizhnik** [comment]
> I have been thinking about a versioned identifier for the layer map, for example, disk_consitent_lsn + number of changes on top, each gc or compaction would be one change, so you'd have lsn.0, after compaction lsn.1, after gc lsn.2. Flushing a new layer would move the lsn to lsn+1, so it would also reset the version to lsn+1.0.

Sorry, I do not understand how it may help to enforce consistency of L1 layers.
Also it is not clear to me how it is possible to move lsn to lsn+1 after flushing.
We already have a lot of mess caused by inclusive/exclusive LSN boundaries, WAL record start/stop  LSNs, LSN alignment and page headers.Proposed LSN+1 will add even more complexity.
And I do not understand which problem you are trying to solve i this way?
Is it just fir testing - to be able to detect when compaction/GC are finished?
If so, it is IMHO overkill.


> So I think we do most of the things you consider here, but we don't have a way to make them visible because of no versioning.

I do not think that we need versioning for read-only layers.
Just atomic fsync+rename will be enough.

**Comment by knizhnik** [comment]
Sorry for false alarm - data corruption is not possible because right ow we actually do `fsync` in `DeltaLayerWriter.finish`. But it actually means that calling `par_fsync::par_fsync` in `compact_level0_phase1` is redundant and can be removed.

I wonder if we can just remove `par_fsync::par_fsync`  from   `compact_level0_phase1` and  `create_delta_layer` or it is better to remove fsync from `DeltaLayerWriter.finish` and change  `par_fsync::par_fsync`  to perform rename itself after fsync completion?

**Comment by koivunej** [comment]
Note: #4265 will change how the fsyncing works around these parts. Previously files and parent directory could had been fsynced in any order, now the files are strictly fsynced before the directory.

**Comment by koivunej** [comment]
Earlier I used lsn+1 to mean moving from a previous disk_consistent_lsn to new disk_consistent_lsn, not doing `lsn_1 = lsn_0 + 1`.

It is possible my thinking is too testing oriented, but I am unsure how we could provide anykind of consistency here except to distinguish the versions better. Safekeepers will not care about our compactions/gc in any case...

**Comment by skyzh** [comment]
So I think this issue is solved given
- the original issue is about L1 file on disk could be corrupted b/c fsync is not done completely
- but we actually do the fsync + rename in file writers
- koivunej mentioned a new issue that fsync is out of order, and now it is fixed

But I believe this is also related to the redundant layer problem, where L1 files are not correctly removed when page server crashes, and it is fixed in https://github.com/neondatabase/neon/pull/4302.

Feel free to reopen if I'm wrong and this is not resolved. Thanks!
